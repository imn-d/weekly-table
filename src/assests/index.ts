export * from './CancelIcon';
export * from './DeleteIcon';
export * from './EditIcon';
export * from './DragIcon';
export * from './CheckIcon';
