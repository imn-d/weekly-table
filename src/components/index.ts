export * from './Background';
export * from './Popup';
export * from './TimeBlockCore';
export * from './CustomTime';
export * from './SchedulerArea';
export * from './SchedulerHeader';
export * from './SchedulerHelper';
export * from './SchedulerLayout';
