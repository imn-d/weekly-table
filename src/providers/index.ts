export * from './CanvasProvider';
export * from './CellsProvider';
export * from './CustomTimeProvider';
export * from './PointerLockProvider';
export * from './PopupProvider';
export * from './Scheduler';
export * from './TimeBlockProvider';
