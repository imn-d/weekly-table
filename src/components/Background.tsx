import React, { memo, useEffect, useRef } from 'react';

import { useCells, useScheduler } from '../providers';

/**
 * Drawing cells background component
 */
export const Background = memo(() => {
  const { helperWidthProp, headerHeightProp, baseZIndex, bottomHeightProp } =
    useScheduler();
  const { widthStep, heightStep, offsetHeight, offsetWidth } = useCells();

  /**
   * Canvas reference
   */
  const canvasRef = useRef<HTMLCanvasElement>(null);

  /**
   * Context reference
   */
  const contextRef = useRef<CanvasRenderingContext2D | null>(null);

  /**
   * Cells redrawing
   */
  useEffect(() => {
    const canvas = canvasRef?.current;
    if (!canvas) return;
    canvas.width = offsetWidth * 2;
    canvas.height = offsetHeight * 2;
    canvas.style.width = `${offsetWidth}px`;
    canvas.style.height = `${offsetHeight}px`;
    const context = canvas.getContext('2d');
    if (!context) return;

    context.scale(2, 2);
    contextRef.current = context;
    contextRef.current.beginPath();
    for (let x = helperWidthProp; x <= offsetWidth; x += widthStep) {
      contextRef.current.moveTo(x, 0);
      contextRef.current.lineTo(x, offsetHeight - bottomHeightProp);
    }
    contextRef.current.strokeStyle = 'rgba(0,0,0, 0.2)';
    contextRef.current.lineWidth = 1;
    contextRef.current.stroke();
    contextRef.current.beginPath();
    for (
      let y = headerHeightProp;
      y <= offsetHeight - bottomHeightProp;
      y += heightStep
    ) {
      contextRef.current.moveTo(helperWidthProp, y);
      contextRef.current.lineTo(offsetWidth, y);
    }
    contextRef.current.strokeStyle = 'rgba(0,0,0, 0.1)';
    contextRef.current.lineWidth = 1;
    contextRef.current.stroke();
    contextRef.current.save();
  }, [
    bottomHeightProp,
    headerHeightProp,
    heightStep,
    helperWidthProp,
    offsetHeight,
    offsetWidth,
    widthStep,
  ]);

  return (
    <canvas
      id={'backgroundCanvas'}
      className={'scheduler-background-canvas'}
      style={{
        position: 'absolute',
        zIndex: baseZIndex + 1,
      }}
      ref={canvasRef}
    />
  );
});
