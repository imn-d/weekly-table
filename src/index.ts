import {
  BlockColorsProps,
  ScheduleGroup,
  SchedulerColumnsProps,
  SchedulerInputProps,
  schedulerColumns,
  schedulerRows,
} from './common';
import Scheduler from './providers/Scheduler';

export default Scheduler;
export { schedulerColumns, schedulerRows };
export type {
  SchedulerColumnsProps,
  SchedulerInputProps,
  ScheduleGroup,
  BlockColorsProps,
};
