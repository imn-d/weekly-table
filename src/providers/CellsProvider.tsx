import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';

import { Cell, CellsProps, HOUR_24 } from '../common';
import { useScheduler } from './Scheduler';

const CellsContext = createContext<CellsProps>({} as CellsProps);
/**
 *  Virtual cells provider context hook
 *  @returns [CellsProps]{@link CellsProps}
 */
export const useCells = (): CellsProps => useContext(CellsContext);

/**
 * Virtual cells provider
 */
const CellsProvider: FC = ({ children }) => {
  const {
    parentRef,
    timeframe,
    columns,
    helperWidthProp,
    headerHeightProp,
    bottomHeightProp,
  } = useScheduler();

  /**
   * Main context state, setting by useEffect
   */
  const [cellsProps, setCellsProps] = useState<CellsProps>({} as CellsProps);

  /**
   * Main context updater
   */
  const setCells = useCallback(
    (update: Cell[]) => {
      setCellsProps({ ...cellsProps, cells: update });
    },
    [cellsProps],
  );

  /**
   * Ms count per timeframe
   */
  const msTime = useMemo(() => {
    return timeframe * 60 * 1000;
  }, [timeframe]);

  /**
   * Calculates context values and build virtual cells area
   */
  useEffect(() => {
    const parent = parentRef?.current;
    if (!parent) return;
    const widthStep = (parent.offsetWidth - helperWidthProp) / columns?.length;
    const areaHeight =
      parent.offsetHeight - headerHeightProp - bottomHeightProp;
    const heightStep = areaHeight / (1440 / timeframe);
    const millisPerPixel = HOUR_24 / areaHeight;
    const maxCellIndex = areaHeight / heightStep - 1;
    const cells = new Array<Cell>();
    let widthPos = helperWidthProp;
    for (let i = 0; i < columns?.length; i++) {
      let heightPos = headerHeightProp;
      for (let j = 0; j < areaHeight / heightStep; j++) {
        cells.push({
          isSelected: false,
          column: i,
          row: j,
          position: {
            x: widthPos,
            y: heightPos,
            w: widthPos + widthStep,
            h: heightPos + heightStep,
          },
        } as Cell);
        heightPos += heightStep;
      }
      widthPos += widthStep;
    }

    setCellsProps({
      widthStep,
      heightStep,
      millisPerPixel,
      cells,
      maxCellIndex,
      offsetWidth: parent.offsetWidth,
      offsetHeight: parent.offsetHeight,
    } as CellsProps);
  }, [
    parentRef?.current?.offsetWidth,
    parentRef?.current?.offsetHeight,
    columns?.length,
    bottomHeightProp,
    headerHeightProp,
    helperWidthProp,
    timeframe,
    parentRef,
  ]);

  return (
    <CellsContext.Provider
      value={{
        ...cellsProps,
        setCells: setCells,
        msTime,
      }}
    >
      {children}
    </CellsContext.Provider>
  );
};

export default CellsProvider;
