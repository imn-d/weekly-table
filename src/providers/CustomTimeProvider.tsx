import React, {
  FC,
  createContext,
  useCallback,
  useContext,
  useState,
} from 'react';

import {
  CustomTimeProps,
  DisplayTime,
  HOUR_24,
  Position,
  TIME,
  Time,
  TimeBlock,
} from '../common';
import { useCells } from './CellsProvider';
import { useTimeBlock } from './TimeBlockProvider';

const CustomTimeContext = createContext<CustomTimeProps>({} as CustomTimeProps);

/**
 *  Custom time provider context hook
 *  @returns [CustomTimeProps]{@link CustomTimeProps}
 */
export const useCustomTime = (): CustomTimeProps =>
  useContext(CustomTimeContext);

/**
 * Custom time provider, used for setting time with second precision
 */
const CustomTimeProvider: FC = ({ children }) => {
  const { millisPerPixel } = useCells();
  const { blocks, setBlocks } = useTimeBlock();

  /**
   * Ms time state
   */
  const [time, setTime] = useState<Time>({} as Time);

  /**
   * Timers relative coordinates
   */
  const [position, setPosition] = useState<Position>();

  /**
   * Editing state
   * If active, other operations with timeblocks will be ignored
   */
  const [isEditing, setEditing] = useState<boolean>(false);

  /**
   * Prime timeblock state
   */
  const [oldBlock, setOldBlock] = useState<TimeBlock>();

  /**
   * Error state with description
   */
  const [isError, setError] = useState<string | undefined>(undefined);

  /**
   * Timers visible time
   */
  const [displayTime, setDisplayTime] = useState<DisplayTime>({
    startTime: '00:00:00',
    endTime: '00:00:00',
  } as DisplayTime);

  /**
   * Extracting timeblock properties and saving it
   */
  const handleObject = useCallback((block: TimeBlock) => {
    setOldBlock(block);
    setPosition({
      top: block.height > 20 ? block.height + block.top + 5 : block.top + 25,
      left: block.left,
    });
    setTime({
      startTime: block.startTime,
      endTime: block.endTime,
    });
    setDisplayTime({
      startTime: new Date(block.startTime).toISOString().substr(11, 8),
      endTime: new Date(block.endTime).toISOString().substr(11, 8),
    });
  }, []);

  /**
   * Timers data extraction
   */
  const handleTime = useCallback(
    (newTime: number, type: boolean) => {
      if (newTime === undefined || newTime === null) {
        setError('Time is null');
        return;
      }
      if (isNaN(newTime)) {
        newTime = 0;
      }
      setError(undefined);
      if (type === TIME.START) {
        setTime({ ...time, startTime: newTime });
        setDisplayTime({
          ...displayTime,
          startTime: new Date(newTime).toISOString().substr(11, 8),
        });
      } else {
        setTime({ ...time, endTime: newTime === 0 ? HOUR_24 : newTime });
        setDisplayTime({
          ...displayTime,
          endTime: new Date(newTime).toISOString().substr(11, 8),
        });
      }
    },
    [displayTime, time],
  );

  /**
   * Timers data reset
   */
  const resetTime = useCallback(() => {
    if (!oldBlock) return;
    setTime({
      startTime: oldBlock.startTime,
      endTime: oldBlock.endTime,
    });
    setDisplayTime({
      startTime: new Date(oldBlock.startTime).toISOString().substr(11, 8),
      endTime: new Date(oldBlock.endTime).toISOString().substr(11, 8),
    });
    setError(undefined);
  }, [oldBlock]);

  /**
   * Updating time to timeblock
   */
  const updateBlock = useCallback(() => {
    if (!oldBlock) return true;
    const filteredBlocks = blocks.filter((b) => b.id !== oldBlock.id);

    if (time.endTime - time.startTime < 1000 * 60) {
      setError('Range < 1 min');
      return true;
    }
    const topY = -(oldBlock.startTime - time.startTime) / millisPerPixel;
    const botY = -(oldBlock.endTime - time.endTime) / millisPerPixel;

    const newBlock = {
      ...oldBlock,
      startTime: time.startTime,
      endTime: time.endTime,
      realStartTime: time.startTime,
      realEndTime: time.endTime,
      top: oldBlock.top + topY,
      height: oldBlock.height + botY - topY,
    };
    setBlocks(filteredBlocks.concat(newBlock));
    return false;
    //eslint-disable-next-line
  }, [blocks, millisPerPixel, oldBlock, time.endTime, time.startTime]);

  return (
    <CustomTimeContext.Provider
      value={{
        isEditing,
        resetTime,
        setEditing,
        handleTime,
        updateBlock,
        position,
        handleObject,
        displayTime,
        isError,
      }}
    >
      {children}
    </CustomTimeContext.Provider>
  );
};

export default CustomTimeProvider;
