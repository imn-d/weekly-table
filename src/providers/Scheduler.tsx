import React, {
  FC,
  createContext,
  useContext,
  useEffect,
  useState,
} from 'react';

import {
  BlockColorsProps,
  SchedulerInputProps,
  SchedulerProps,
  schedulerColumns,
  schedulerRows,
} from '../common';
import { Background, CustomTime, Popup, SchedulerLayout } from '../components';
import CanvasProvider from './CanvasProvider';
import CellsProvider from './CellsProvider';
import CustomTimeProvider from './CustomTimeProvider';
import PointerLockProvider from './PointerLockProvider';
import PopupProvider from './PopupProvider';
import TimeBlockProvider from './TimeBlockProvider';

const SchedulerContext = createContext<SchedulerProps>({} as SchedulerProps);
/**
 *  Scheduler provider context hook
 *  @returns [SchedulerProps]{@link SchedulerProps}
 */
export const useScheduler = (): SchedulerProps => useContext(SchedulerContext);

/**
 * Root scheduler components
 * @param props{@link SchedulerInputProps} initial props
 */
const Scheduler: FC<SchedulerInputProps> = (props: SchedulerInputProps) => {
  const [update, setUpdate] = useState<number>(0);

  useEffect(() => {
    window.addEventListener('resize', () => setUpdate((update) => update + 1));
    return () => {
      window.removeEventListener('resize', () =>
        setUpdate((update) => update + 1),
      );
    };
  }, []);

  return (
    <SchedulerContext.Provider
      value={{
        parentRef: props.parentRef,
        blockColors: props?.blockColors
          ? props.blockColors
          : ({
              common: '#ff5722',
              temp: '#c6a700',
              draw: '#ff8a50',
              hover: '#ff3d00',
            } as BlockColorsProps),
        headerHeightProp: props?.headerHeightProp ? props.headerHeightProp : 80,
        helperWidthProp: props?.helperWidthProp ? props.helperWidthProp : 80,
        bottomHeightProp: props?.bottomHeightProp ? props.bottomHeightProp : 20,
        baseZIndex: props?.baseZIndex ? props.baseZIndex : 0,
        timeframe: props?.timeframe ? props.timeframe : 60,
        columns: props?.columns ? props.columns : schedulerColumns,
        rows: props?.rows ? props.rows : schedulerRows,
        requiredTZOffset: props?.requiredTZOffset,
        mouseSpeed: props?.mouseSpeed ? props.mouseSpeed : 3,
        defaultValue: props?.defaultValue,
        onChange: props?.onChange,
      }}
    >
      <CellsProvider>
        <CanvasProvider>
          <PointerLockProvider>
            <TimeBlockProvider>
              <PopupProvider>
                <CustomTimeProvider>
                  <Background />
                  <SchedulerLayout />
                  <Popup />
                  <CustomTime />
                </CustomTimeProvider>
              </PopupProvider>
            </TimeBlockProvider>
          </PointerLockProvider>
        </CanvasProvider>
      </CellsProvider>
    </SchedulerContext.Provider>
  );
};

export default Scheduler;
