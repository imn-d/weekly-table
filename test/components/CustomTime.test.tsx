import { mount } from 'enzyme';
import React from 'react';
import { act } from 'react-dom/test-utils';

import {
  CellsProps,
  CustomTimeProps,
  DisplayTime,
  Position,
  SchedulerProps,
} from '../../src/common';
import { CustomTime } from '../../src/components';
import * as uc from '../../src/providers/CellsProvider';
import * as ut from '../../src/providers/CustomTimeProvider';
import * as us from '../../src/providers/Scheduler';

describe('CustomTime Component', () => {
  const useSchedulerMock = jest.spyOn(us, 'useScheduler');
  const useCellMock = jest.spyOn(uc, 'useCells');
  const useCustomMock = jest.spyOn(ut, 'useCustomTime');

  const customMock = {
    isEditing: true,
    position: { top: 0, left: 0 } as Position,
    displayTime: {
      startTime: '00:00:00',
      endTime: '18:00:00',
    } as DisplayTime,
    handleTime: (newTime, type) => console.log(newTime, type),
    isError: undefined,
  } as CustomTimeProps;

  beforeEach(() => {
    jest.clearAllMocks();

    useSchedulerMock.mockReturnValue({
      baseZIndex: 0,
    } as SchedulerProps);

    useCellMock.mockReturnValue({
      widthStep: 20,
    } as CellsProps);
  });

  it('Base Render; no Error', () => {
    useCustomMock.mockReturnValueOnce(customMock);

    const wrapper = mount(<CustomTime />);

    expect(wrapper.find('#customTimeError').hostNodes()).toHaveLength(0);

    expect(useSchedulerMock).toBeCalledTimes(1);
    expect(useCellMock).toBeCalledTimes(1);
    expect(useCustomMock).toBeCalledTimes(1);

    expect(wrapper).toMatchSnapshot('baseRender');
    wrapper.unmount();
  });

  it('Display Error Test', () => {
    useCustomMock.mockReturnValueOnce({ ...customMock, isError: 'MY_ERROR' });

    const wrapper = mount(<CustomTime />);
    const errorSpan = wrapper.find('#customTimeError').hostNodes();
    expect(errorSpan).toHaveLength(1);
    expect(errorSpan).toMatchInlineSnapshot(`
      <span
        className="custom-time-error"
        id="customTimeError"
        style={
          Object {
            "position": "absolute",
            "top": 55,
            "width": 20,
          }
        }
      >
        MY_ERROR
      </span>
    `);
    wrapper.unmount();
  });

  it('Handle Time Test', async () => {
    const spyHandler = jest.fn().mockImplementation(() => Promise.resolve());
    useCustomMock.mockReturnValueOnce({
      ...customMock,
      handleTime: spyHandler,
    });

    const wrapper = mount(<CustomTime />);

    const startInput = wrapper.find('#customTimeInputStart').hostNodes();
    const endInput = wrapper.find('#customTimeInputEnd').hostNodes();

    expect(startInput).toHaveLength(1);
    expect(endInput).toHaveLength(1);

    expect(spyHandler).toBeCalledTimes(0);

    await act(async () => {
      startInput.simulate('change', {
        target: { valueAsNumber: 900000, id: 'customTimeInputStart' },
      });
      expect(spyHandler).toBeCalledWith(900000, true);
      endInput.simulate('change', {
        target: { valueAsNumber: 1500000, id: 'customTimeInputEnd' },
      });
      expect(spyHandler).toBeCalledWith(1500000, false);
    });
    expect(spyHandler).toBeCalledTimes(2);
    wrapper.unmount();
  });

  it('Display Test', async () => {
    useCustomMock.mockReturnValueOnce(customMock);

    let wrapper = mount(<CustomTime />);
    let parent = wrapper.find('#customTimeWrapper').hostNodes();
    expect(parent.props()?.style?.display).toBe('flex');

    wrapper.unmount();

    useCustomMock.mockReturnValueOnce({ ...customMock, isEditing: false });

    wrapper = mount(<CustomTime />);
    parent = wrapper.find('#customTimeWrapper').hostNodes();
    expect(parent.props()?.style?.display).toBe('none');

    wrapper.unmount();
  });
});
