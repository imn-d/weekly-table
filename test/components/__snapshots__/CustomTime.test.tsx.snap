// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomTime Component Base Render; no Error: baseRender 1`] = `
<CustomTime>
  <div
    className="scheduler-text custom-time-wrapper"
    id="customTimeWrapper"
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "left": 0,
        "maxWidth": 15,
        "position": "absolute",
        "top": 0,
        "width": 15,
        "zIndex": 6,
      }
    }
  >
    <input
      className="custom-time-input custom-time-input-start"
      id="customTimeInputStart"
      onChange={[Function]}
      step={1}
      style={
        Object {
          "maxWidth": 15,
          "width": 15,
        }
      }
      type="time"
      value="00:00:00"
    />
    <input
      className="custom-time-input custom-time-input-end"
      id="customTimeInputEnd"
      onChange={[Function]}
      step={1}
      style={
        Object {
          "maxWidth": 15,
          "width": 15,
        }
      }
      type="time"
      value="18:00:00"
    />
  </div>
</CustomTime>
`;
